--- ilorig/Daemaged.ManagedHell.il	2010-07-28 16:55:32.892148400 +0300
+++ il/Daemaged.ManagedHell.il	2010-07-28 16:52:46.779761300 +0300
@@ -4053,21 +4053,11 @@
     {
       // Code size       38 (0x26)
       .maxstack  2
-      .locals init ([0] !T poof,
-               [1] native int marker,
-               [2] native int* pmarker)
+      .locals init ([0] !T poof)
       IL_0000:  ldloca.s   poof
-      IL_0002:  initobj    !T
-      IL_0008:  ldloca.s   marker
-      IL_000a:  conv.u
-      IL_000b:  sizeof     [mscorlib]System.IntPtr
-      IL_0011:  sub
-      IL_0012:  stloc.2
-      IL_0013:  ldloc.2
       IL_0014:  ldarg.0
       IL_0015:  ldfld      uint8* class Daemaged.ManagedHell.BluePill`1/BluePillEnumerator<!T>::_p
-      IL_001a:  call       native int [mscorlib]System.IntPtr::op_Explicit(void*)
-      IL_001f:  stobj      [mscorlib]System.IntPtr
+      IL_001f:  stobj      uint8*
       IL_0024:  ldloc.0
       IL_0025:  ret
     } // end of method BluePillEnumerator::get_Current
@@ -4246,26 +4236,15 @@
     {
       // Code size       50 (0x32)
       .maxstack  4
-      .locals init ([0] !T poof,
-               [1] native int marker,
-               [2] native int* pmarker)
+      .locals init ([0] !T poof)
       IL_0000:  ldloca.s   poof
-      IL_0002:  initobj    !T
-      IL_0008:  ldloca.s   marker
-      IL_000a:  conv.u
-      IL_000b:  sizeof     [mscorlib]System.IntPtr
-      IL_0011:  sub
-      IL_0012:  stloc.2
-      IL_0013:  ldloc.2
       IL_0014:  ldarg.0
       IL_0015:  ldfld      void* class Daemaged.ManagedHell.BluePill`1/BluePillList<!T>::_p
-      IL_001a:  call       native int [mscorlib]System.IntPtr::op_Explicit(void*)
       IL_001f:  ldarg.1
       IL_0020:  ldsfld     int32 class Daemaged.ManagedHell.BluePill`1<!T>::_size
       IL_0025:  mul
-      IL_0026:  call       native int [mscorlib]System.IntPtr::op_Addition(native int,
-                                                                           int32)
-      IL_002b:  stobj      [mscorlib]System.IntPtr
+      IL_0026:  add
+      IL_002b:  stobj      uint8*
       IL_0030:  ldloc.0
       IL_0031:  ret
     } // end of method BluePillList::get_Item
@@ -4400,19 +4379,13 @@
              [1] native int** pmarker)
     IL_0000:  ldsfld     class [mscorlib]System.Type class Daemaged.ManagedHell.BluePill`1<!T>::_type
     IL_0005:  callvirt   instance bool [mscorlib]System.Type::get_IsValueType()
-    IL_000a:  brfalse.s  IL_0017
+    IL_000a:  brfalse.s  IL_0023
 
     IL_000c:  ldstr      "Cannot cast a pointer to an object into an array"
     IL_0011:  newobj     instance void [mscorlib]System.ArgumentException::.ctor(string)
     IL_0016:  throw
-
-    IL_0017:  ldloca.s   marker
-    IL_0019:  conv.u
-    IL_001a:  sizeof     native int*
-    IL_0020:  sub
-    IL_0021:  stloc.1
-    IL_0022:  ldloc.1
-    IL_0023:  ldind.i
+    
+    IL_0023:  ldarga.s 0
     IL_0024:  ret
   } // end of method BluePill`1::ToPointer
 
@@ -4420,41 +4393,26 @@
   {
     // Code size       87 (0x57)
     .maxstack  2
-    .locals init ([0] !T poof,
-             [1] native int marker,
-             [2] native int* rtthPtr,
-             [3] native int* pmarker)
+    .locals init ([0] !T poof)
     IL_0000:  ldsfld     class [mscorlib]System.Type class Daemaged.ManagedHell.BluePill`1<!T>::_type
     IL_0005:  callvirt   instance bool [mscorlib]System.Type::get_IsArray()
-    IL_000a:  brfalse.s  IL_0017
+    IL_000a:  brfalse.s  IL_001f
 
     IL_000c:  ldstr      "Cannot cast a pointer to an object into an array"
     IL_0011:  newobj     instance void [mscorlib]System.ArgumentException::.ctor(string)
     IL_0016:  throw
 
-    IL_0017:  ldloca.s   poof
-    IL_0019:  initobj    !T
     IL_001f:  ldarg.0
-    IL_0020:  stloc.2
-    IL_0021:  ldloc.2
-    IL_0022:  ldobj      [mscorlib]System.IntPtr
+    IL_0022:  ldobj      native int
     IL_0027:  ldsfld     native int class Daemaged.ManagedHell.BluePill`1<!T>::_rtth
-    IL_002c:  call       bool [mscorlib]System.IntPtr::op_Inequality(native int,
-                                                                     native int)
-    IL_0031:  brfalse.s  IL_003e
+    IL_0031:  beq.s  IL_003e
 
-    IL_0033:  ldloc.2
+    IL_0033:  ldarg.0
     IL_0034:  ldsfld     native int class Daemaged.ManagedHell.BluePill`1<!T>::_rtth
-    IL_0039:  stobj      [mscorlib]System.IntPtr
-    IL_003e:  ldloca.s   marker
-    IL_0040:  conv.u
-    IL_0041:  sizeof     [mscorlib]System.IntPtr
-    IL_0047:  sub
-    IL_0048:  stloc.3
-    IL_0049:  ldloc.3
-    IL_004a:  ldloc.2
-    IL_004b:  call       native int [mscorlib]System.IntPtr::op_Explicit(void*)
-    IL_0050:  stobj      [mscorlib]System.IntPtr
+    IL_0039:  stobj      native int
+    IL_003e:  ldloca.s   poof
+              ldarg.0
+    IL_0050:  stobj      void *
     IL_0055:  ldloc.0
     IL_0056:  ret
   } // end of method BluePill`1::FromPointer
@@ -4471,42 +4429,28 @@
              [4] native int* pmarker)
     IL_0000:  ldsfld     class [mscorlib]System.Type class Daemaged.ManagedHell.BluePill`1<!T>::_type
     IL_0005:  callvirt   instance bool [mscorlib]System.Type::get_IsArray()
-    IL_000a:  brtrue.s   IL_0017
+    IL_000a:  brtrue.s   IL_001f
 
     IL_000c:  ldstr      "Cannot cast a pointer to an array into an object"
     IL_0011:  newobj     instance void [mscorlib]System.ArgumentException::.ctor(string)
     IL_0016:  throw
 
-    IL_0017:  ldloca.s   poof
-    IL_0019:  initobj    !T
     IL_001f:  ldarg.0
-    IL_0020:  sizeof     [mscorlib]System.IntPtr
+    IL_0020:  sizeof     native int
     IL_0026:  add
-    IL_0027:  stloc.2
-    IL_0028:  ldloc.2
     IL_0029:  ldarg.1
-    IL_002a:  stobj      [mscorlib]System.IntPtr
+    IL_002a:  stobj      native int
     IL_002f:  ldarg.0
-    IL_0030:  stloc.3
-    IL_0031:  ldloc.3
-    IL_0032:  ldobj      [mscorlib]System.IntPtr
+    IL_0032:  ldobj      native int
     IL_0037:  ldsfld     native int class Daemaged.ManagedHell.BluePill`1<!T>::_rtth
-    IL_003c:  call       bool [mscorlib]System.IntPtr::op_Inequality(native int,
-                                                                     native int)
-    IL_0041:  brfalse.s  IL_004e
+    IL_0041:  beq.s  IL_004e
 
-    IL_0043:  ldloc.3
+    IL_0043:  ldarg.0
     IL_0044:  ldsfld     native int class Daemaged.ManagedHell.BluePill`1<!T>::_rtth
-    IL_0049:  stobj      [mscorlib]System.IntPtr
-    IL_004e:  ldloca.s   marker
-    IL_0050:  conv.u
-    IL_0051:  sizeof     [mscorlib]System.IntPtr
-    IL_0057:  sub
-    IL_0058:  stloc.s    pmarker
-    IL_005a:  ldloc.s    pmarker
-    IL_005c:  ldloc.3
-    IL_005d:  call       native int [mscorlib]System.IntPtr::op_Explicit(void*)
-    IL_0062:  stobj      [mscorlib]System.IntPtr
+    IL_0049:  stobj      native int
+    IL_004e:  ldloca.s   poof
+              ldarg.0
+    IL_0062:  stobj      void *
     IL_0067:  ldloc.0
     IL_0068:  ret
   } // end of method BluePill`1::FromPointer
